In [35]:
HTML(read(open("style.css"), String))

HTML{String}("<link href='http://fonts.googleapis.com/css?family=Alegreya+Sans:100,300,400,500,700,800,900,100italic,300italic,400italic,500italic,700italic,800italic,900italic' rel='stylesheet' type='text/css'>\r\n<link href='http://fonts.googleapis.com/css?family=Arvo:400,700,400italic' rel='stylesheet' type='text/css'>\r\n<link href='http://fonts.googleapis.com/css?family=PT+Mono' rel='stylesheet' type='text/css'>\r\n<link href='http://fonts.googleapis.com/css?family=Shadows+Into+Light' rel='stylesheet' type='text/css'>\r\n<link href='http://fonts.googleapis.com/css?family=Philosopher:400,700,400italic,700italic' rel='stylesheet' type='text/css'>\r\n\r\n<style>\r\n\r\n@font-face {\r\n    font-family: \"Computer Modern\";\r\n    src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');\r\n}\r\n\r\n.container { width: 100% }\r\n\r\n/* Formatting for header cells */\r\n.text_cell_render h1 {\r\n    font-family: 'Philosopher', sans-serif;\r\n    font-weight: 400;\r\n    font-size: 2.2em;\r\n    line-height: 100%;\r\n    color: rgb(0, 80, 120);\r\n    margin-bottom: 0.1em;\r\n    margin-top: 0.1em;\r\n    display: block;\r\n}\t\r\n.text_cell_render h2 {\r\n    font-family: 'Philosopher', serif;\r\n    font-weight: 400;\r\n    font-size: 1.9em;\r\n    line-height: 100%;\r\n    color: rgb(200,100,0);\r\n    margin-bottom: 0.1em;\r\n    margin-top: 0.1em;\r\n    display: block;\r\n}\t\r\n\r\n.text_cell_render h3 {\r\n    font-family: 'Philosopher', serif;\r\n    margin-top:12px;\r\n    margin-bottom: 3px;\r\n    font-style: italic;\r\n    color: rgb(94,127,192);\r\n}\r\n\r\n.text_cell_render h4 {\r\n    font-family: 'Philosopher', serif;\r\n}\r\n\r\n.text_cell_render h5 {\r\n    font-family: 'Alegreya Sans', sans-serif;\r\n    font-weight: 300;\r\n    font-size: 16pt;\r\n    color: grey;\r\n    font-style: italic;\r\n    margin-bottom: .1em;\r\n    margin-top: 0.1em;\r\n    display: block;\r\n}\r\n\r\n.text_cell_render h6 {\r\n    font-family: 'PT Mono', sans-serif;\r\n    font-weight: 300;\r\n    font-size: 10pt;\r\n    color: grey;\r\n    margin-bottom: 1px;\r\n    margin-top: 1px;\r\n}\r\n\r\n.text_cell_render em {\r\n    font-family: 'Philosopher', sans-serif;\r\n    color:        blue;\r\n    background-color: rgb(255,220,180);\r\n    font-size:    110%;\r\n    margin-left:   2px;\r\n    margin-right:  2px;\r\n    font-weight:   100;\r\n}\r\n\r\n.text_cell_render b {\r\n    color:            rgb(255,195,195);\r\n    background-color: rgb(0,0,0);\r\n    font-size:    110%;\r\n    margin-left:   2px;\r\n    margin-right:  2px;\r\n    font-weight:   650;\r\n}\r\n\r\n.text_cell_render tt {\r\n    font-size:    120%;\r\n    margin-left:   2px;\r\n    margin-right:  2px;\r\n    font-weight:   150;\r\n}\r\n\r\n.Codemirror {\r\n    font-family: \"PT Mono\";\r\n    font-size: 100%;\r\n}\r\n\r\n</style>\r\n\r\n")

# Quiescence Search Testing

In [36]:
using Pkg
# Pkg.add("NBInclude")
using NBInclude
# Pkg.add("BenchmarkTools")
using BenchmarkTools

In [37]:
@nbinclude("QuiescenceSearch.ipynb")

strategicQS (generic function with 1 method)

In [38]:
@nbinclude("IterativeDeepening.ipynb")

iterativeDeepening (generic function with 1 method)

In [39]:
zobrist = generate_zobrist_hashing()

ZobristHashing(UInt64[0x872051a8e710005c 0xeb9ec60793fcc70f … 0xaffd85137b219094 0x9826309b2af46a00; 0x2f010468282057e0 0xea629a2b4bc3486a … 0x7b3e3a89e8af9775 0x6f0075901eccdc56; … ; 0x17c633d50778192d 0x89a01832c96880da … 0x6530ec7d8a1a4888 0xc77c6ea88071bf68; 0x020b7456c8123dd4 0xd847f3652678d965 … 0x6088e4674a12481b 0x71da61d416eae808], UInt64[0x90e48f1bcf8fee61, 0x0699bc148126afee, 0x711c44611f421a59, 0xac1c5a017b872d66], UInt64[0xe947377e4c4b06ea, 0x113412569c528f4c, 0xfaa01e62ddb26e54, 0xb1ef51bb9e1844c5, 0x002512227c2f7448, 0x98c3f268ccaa9b25, 0xd8c3b694dfc4696d, 0xaa889408956eb389], 0x74673e0c4c0a39e7)

## Benchmarking tests

In [40]:
gCache = Dict()
b2 = fromfen("k7/pp5r/7r/7r/7R/7R/PP5R/K7 w - - 0 1")
display(b2)
@time begin
    # @btime noQuiesce(b2, evaluate_position(b2))
    # quiesceSee(b2, evaluate_position(b2), -Inf, Inf)
    # quiesceNoPrune(b2, evaluate_position(b2), zobrist_hash(b2))
    # quiesce(b2, evaluate_position(b2), zobrist_hash(b2), 0, -Inf, Inf)
    # quiesceMax(b2, evaluate_position(b2), zobrist_hash(b2), 0, -Inf, Inf)
end

Board (k7/pp5r/7r/7r/7R/7R/PP5R/K7 w - -):
 k  -  -  -  -  -  -  - 
 p  p  -  -  -  -  -  r 
 -  -  -  -  -  -  -  r 
 -  -  -  -  -  -  -  r 
 -  -  -  -  -  -  -  R 
 -  -  -  -  -  -  -  R 
 P  P  -  -  -  -  -  R 
 K  -  -  -  -  -  -  -

  0.000001 seconds


In [41]:
gCache

Dict{Any, Any}()

In [42]:
gCache = Dict()
b3 = fromfen("rn3rk1/pbppq1pp/1p2pb2/4N2Q/3PN3/3B4/PPP2PPP/R3K2R w KQ - 7 11")
display(b3)

@time begin
    # @btime noQuiesce(b3, evaluate_position(b3))
    # @btime quiesceSee(b3, evaluate_position(b3), -Inf, Inf)
    # quiesceNoPrune(b3, evaluate_position(b3), zobrist_hash(b3))
    # quiesce(b3, evaluate_position(b3), zobrist_hash(b3), 0, -Inf, Inf)
    quiesceMax(b3, evaluate_position(b3), zobrist_hash(b3), 0, -Inf, Inf)
end

Board (rn3rk1/pbppq1pp/1p2pb2/4N2Q/3PN3/3B4/PPP2PPP/R3K2R w KQ -):
 r  n  -  -  -  r  k  - 
 p  b  p  p  q  -  p  p 
 -  p  -  -  p  b  -  - 
 -  -  -  -  N  -  -  Q 
 -  -  -  P  N  -  -  - 
 -  -  -  B  -  -  -  - 
 P  P  P  -  -  P  P  P 
 R  -  -  -  K  -  -  R

  0.264450 seconds (401.11 k allocations: 18.879 MiB, 4.05% gc time, 99.56% compilation time)


120

Benchmark results of b3 = fromfen("rn3rk1/pbppq1pp/1p2pb2/4N2Q/3PN3/3B4/PPP2PPP/R3K2R w KQ - 7 11")
1. noQuiesce   21.700 μs (316 allocations: 25.88 KiB) => -40
1. quiesceSee  27.000 μs (391 allocations: 43.20 KiB) => -40
1. quiesce     52.700 μs (692 allocations: 45.09 KiB) => -40
1. quiesce no Mem  49.100 μs (656 allocations: 44.15 KiB) => -40

In [43]:
gCache = Dict()
b4 = fromfen("r1b1k2r/pp1nqpp1/4p2p/3pP1N1/8/3BQ3/PP3PPP/2R2RK1 w kq - 0 1")
display(b4)

@time begin
    # @btime noQuiesce(b4, evaluate_position(b4))
    # @btime quiesceSee(b4, evaluate_position(b4), -Inf, Inf)
    # quiesce(b4, evaluate_position(b4), zobrist_hash(b4), 0, -Inf, Inf)
    quiesceMax(b4, evaluate_position(b4), zobrist_hash(b4), 0, -Inf, Inf)
end

Board (r1b1k2r/pp1nqpp1/4p2p/3pP1N1/8/3BQ3/PP3PPP/2R2RK1 w kq -):
 r  -  b  -  k  -  -  r 
 p  p  -  n  q  p  p  - 
 -  -  -  -  p  -  -  p 
 -  -  -  p  P  -  N  - 
 -  -  -  -  -  -  -  - 
 -  -  -  B  Q  -  -  - 
 P  P  -  -  -  P  P  P 
 -  -  R  -  -  R  K  -

  0.001773 seconds (10.56 k allocations: 789.102 KiB)


-40

In [47]:
gCache = Dict()
b5 = fromfen("8/4k3/8/4p3/8/2BK4/8/q7 w - - 0 1")
display(b5)

@time begin
    # quiesceNoPrune(b5, evaluate_position(b5), zobrist_hash(b5))
    # quiesce(b5, evaluate_position(b5), zobrist_hash(b5), 0, -Inf, Inf)
    quiesceMax(b5, evaluate_position(b5), zobrist_hash(b5), 0, -Inf, Inf)
end

Board (8/4k3/8/4p3/8/2BK4/8/q7 w - -):
 -  -  -  -  -  -  -  - 
 -  -  -  -  k  -  -  - 
 -  -  -  -  -  -  -  - 
 -  -  -  -  p  -  -  - 
 -  -  -  -  -  -  -  - 
 -  -  B  K  -  -  -  - 
 -  -  -  -  -  -  -  - 
 q  -  -  -  -  -  -  -

  0.000169 seconds (319 allocations: 23.492 KiB)


215

Benchmark results:
1. noQuiescece 0.027139 seconds (22.88 k allocations: 1.142 MiB, 98.74% compilation time) ==> 215
1. quiesce 0.139313 seconds (142.19 k allocations: 7.917 MiB, 85.17% compilation time) ==> 640
1. quiesceMax 0.000169 seconds (319 allocations: 23.492 KiB) ==> 215

In [45]:
b1 = fromfen("rn3rk1/pbppq1pp/1p2pb2/4N2Q/3PN3/3B4/PPP2PPP/R3K2R w KQ - 7 11")
display(b1)
@btime see(b1, movefromstring("h5e8"))  #, movefromstring("h5e8")

Board (rn3rk1/pbppq1pp/1p2pb2/4N2Q/3PN3/3B4/PPP2PPP/R3K2R w KQ -):
 r  n  -  -  -  r  k  - 
 p  b  p  p  q  -  p  p 
 -  p  -  -  p  b  -  - 
 -  -  -  -  N  -  -  Q 
 -  -  -  P  N  -  -  - 
 -  -  -  B  -  -  -  - 
 P  P  P  -  -  P  P  P 
 R  -  -  -  K  -  -  R

  235.492 ns (4 allocations: 560 bytes)


-10

Benchmark results

1. chess.jl see function 241.727 ns
1. isTactical 213.452 ns
1. isQuiet 1610.000 ns
    1. 345ns durch moves(State)